### PR TITLE
[3.8] AQL function SHARD_ID

### PR DESCRIPTION
**Pull request by kvahed**
### Scope & Purpose

*Added new AQL function, `SHARD_ID(<col>, {shardKeys...})`, which obtains the shard to a set of shard keys, `shardKeys...` in collection `col`*

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [X] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] Backport for 3.9: *(Please link PR)*
- [ ] Backport for 3.8: *(Please link PR)*
- [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -8440,6 +8440,61 @@ AqlValue Functions::DateFormat(ExpressionContext* expressionContext,
...
+
+  if (cluster) {
+    auto& ci = vocbase.server().getFeature<ClusterFeature>().clusterInfo();
+    collection = ci.getCollection(dbName, colName);
+  } else { // single server, agents, active failover try to not break cluster ready code
+    methods::Collections::lookup(vocbase, colName, collection);
+  }
+
+  if (collection == nullptr) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
```
Comment (gnusi) [inline arangod/Aql/Functions.cpp:8479]: by convention AQL functions normally don't throw exception, but return `null` + register warning instead

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -8440,6 +8440,61 @@ AqlValue Functions::DateFormat(ExpressionContext* expressionContext,
...
+      TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND, "could not find collection: " + colName + " in database " + dbName);
+  }
+
+  std::string shardId;
+  if (cluster) {
+    auto const errorCode = collection->getResponsibleShard(keys, true, shardId);
+    if(errorCode != TRI_ERROR_NO_ERROR) {
+      THROW_ARANGO_EXCEPTION_MESSAGE(
```
Comment (gnusi) [inline arangod/Aql/Functions.cpp:8487]: by convention AQL functions normally don't throw exception, but return `null` + register warning instead

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -8440,6 +8440,61 @@ AqlValue Functions::DateFormat(ExpressionContext* expressionContext,
...
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+      TRI_ERROR_BAD_PARAMETER,
+      "second parameter to SHARD_ID must be an object");
+  }
+
+  auto& vocbase = expressionContext->trx().vocbase();
+  auto role = ServerState::instance()->getRole();
+  std::shared_ptr<LogicalCollection> collection;
+  auto const colName = col.copyString();
+  auto const dbName = vocbase.name();
```
Comment (gnusi) [inline arangod/Aql/Functions.cpp:8468]: don't copy `std::string`

File: arangod/Aql/AqlFunctionFeature.cpp [hunk 4]
```diff
@@ -434,6 +434,7 @@ void AqlFunctionFeature::addMiscFunctions() {
   add({"PARSE_IDENTIFIER", ".", flags, &Functions::ParseIdentifier});
   add({"IS_SAME_COLLECTION", ".h,.h", flags, &Functions::IsSameCollection});
   add({"DECODE_REV", ".", flags, &Functions::DecodeRev});
+  add({"SHARD_ID", ".h|.", flags, &Functions::ShardId});
```
Comment (gnusi) [inline arangod/Aql/AqlFunctionFeature.cpp:437]: we must disallow using this function in `AqlAnalyzer`

File: tests/js/server/aql/aql-functions-misc.js [hunk 5]
```diff
@@ -707,6 +707,286 @@ function ahuacatlMiscFunctionsTestSuite () {
...
+        for (i = 0; i < 1000; ++i) {
+          a = Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5)+i;
+          b = Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5)+i;
+          doc = {a, b, i, d: db.cl.insert({a,i,b})};
+          docs.push(doc);
+        }
+        docs.forEach(function (doc) {
+          sid = db._query('RETURN SHARD_ID("cl", {a:@va, b:@vb})', {va:doc.a,vb:doc.b}).toArray();
+          assertEqual(sid.length, 1);
+          d = db._query('FOR j in cl FILTER j.i == @val RETURN j', {val:doc.i}, {shardIds: sid}).toArray();
```
Comment (gnusi) [inline tests/js/server/aql/aql-functions-misc.js:981]: this test must be executed on single-server as well

File: arangod/Aql/Functions.cpp [hunk 6]
```diff
@@ -8440,6 +8440,61 @@ AqlValue Functions::DateFormat(ExpressionContext* expressionContext,
...
+                            VPackFunctionParameters const& parameters) {
+
+  // Expecting 2 parameters
+  // 0 : collection name or id
+  // 1 : {shardKey1 : k1, shardKey2: k2, ..., shardKeyn: kn}
+  auto const col = extractFunctionParameterValue(parameters, 0).slice();
+  auto const keys = extractFunctionParameterValue(parameters, 1).slice();
+
+  if (!col.isString()) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
```
Comment (gnusi) [inline arangod/Aql/Functions.cpp:8454]: by convention AQL functions normally don't throw exception, but return null + register warning instead

File: arangod/Aql/Functions.cpp [hunk 7]
```diff
@@ -8440,6 +8440,61 @@ AqlValue Functions::DateFormat(ExpressionContext* expressionContext,
...
+      TRI_ERROR_BAD_PARAMETER,
+      "first parameter to SHARD_ID must be a string");
+  }
+  if (!keys.isObject()) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
```
Comment (gnusi) [inline arangod/Aql/Functions.cpp:8459]: by convention AQL functions normally don't throw exception, but return null + register warning instead

**Review by gnusi — CHANGES_REQUESTED** [review]

File: arangod/Aql/AqlFunctionFeature.cpp (hunk 4 above)
Comment (kvahed) [reply arangod/Aql/AqlFunctionFeature.cpp:437]: added `FF::CanReadDocuments` to make sure that cannot be used in ANALYZERS

File: arangod/Aql/Functions.cpp (hunk 3 above)
Comment (kvahed) [reply arangod/Aql/Functions.cpp:8468]: `getCollection` doesn't handle `std::string_view` well

**Review by gnusi — APPROVED** [review]

**Pull request merged**